app/controllers/application_controller.rb

```diff
@@ -50,8 +52,22 @@ def decorated_current_user
     @decorated_current_user ||= current_user.decorate
   end
 
+  def decorated_true_user
+    @decorated_true_user ||= true_user.decorate
+  end
+
   def current_user
-    @current_user ||= User.find(session[:user_id]) if session[:user_id]
+    user_id = session[:impersonated_user_id] || session[:user_id]
+
+    if @current_user.present?
+      if @current_user.id != user_id
+        @current_user = User.find(user_id)
+      else
+        @current_user ||= User.find(user_id)
```

Comment:
Why is this line needed? In this case `@current_user.id` is already correct, right?